utils/src/lib.rs

```diff
@@ -62,9 +65,41 @@ pub fn remove_extra_costs(gas: U256, calldata: &[u8]) -> U256 {
     gas - calldata_cost - BASE_TX_COST
 }
 
+/// Flattens a group of contracts into maps of all events and functions
+pub fn flatten_funcs_and_events(
+    contracts: &BTreeMap<String, (Abi, Vec<u8>)>,
+) -> (BTreeMap<[u8; 4], Function>, BTreeMap<H256, Event>, Abi) {
+    let flattened_funcs: BTreeMap<[u8; 4], Function> = contracts
+        .iter()
+        .flat_map(|(_name, (abi, _code))| abi.functions().cloned().collect::<Vec<Function>>())
+        .collect::<Vec<Function>>()
+        .into_iter()
```

Comment:
You should be able to avoid the double collect/clone, this prob is expensive for larger projects

Reply:
Hows this:
```rust
let flattened_funcs: BTreeMap<[u8; 4], Function> = contracts
        .iter()
        .flat_map(|(_name, (abi, _code))| {
            abi.functions().map(|func| {
                (func.short_signature(), func.clone())
            })
            .collect::<BTreeMap<[u8; 4], Function>>()
        })
        .collect();
```
?

Fundamentally, we do sort of need to do 2 collect + flattens imo

Reply:
I was thinking:

```
let flattened_funcs = contracts
        .iter()
        .flat_map(|(_name, (abi, _code))| {
            abi.functions().map(|func| {
                (func.short_signature(), func.clone())
            })
        })
        .flatten()
       .collect::<BTreeMap<[u8; 4], Function>>()
```

Reply:
this does not work:
```
error[E0277]: `([u8; 4], Function)` is not an iterator
    --> utils/src/lib.rs:79:10
     |
79   |         .flatten()
     |          ^^^^^^^ `([u8; 4], Function)` is not an iterator
     |
     = help: the trait `Iterator` is not implemented for `([u8; 4], Function)`
     = note: required because of the requirements on the impl of `IntoIterator` for `([u8; 4], Function)`
```